frontend: Pod: Add prettify option for JSON Logs

userAdityaa opened the pull request:
### PR Description
This pull request introduces a feature that formats JSON logs from pod logs into a more human-readable format. It enhances the user experience by prettifying and syntax-highlighting the logs for better readability. Users also have the option to download the prettified JSON logs.


### Screenshot of the changes: 
1. Toggle button for Prettifying `JSON` logs and Fromat `JSON` values
<img width="1281" alt="Screenshot 2025-05-26 at 3 08 10 AM" src="https://github.com/user-attachments/assets/2923d827-47e9-46bd-8fca-37989eee0d70" />

2. If there are no JSON logs available, the toggle button will not be displayed:
<img width="1279" alt="Screenshot 2025-05-26 at 3 09 05 AM" src="https://github.com/user-attachments/assets/65cb0e6d-3a15-463b-a0b8-d55a0a1bd8f6" />

3. Prettify only:
<img width="1280" alt="Screenshot 2025-05-26 at 3 10 29 AM" src="https://github.com/user-attachments/assets/85b98708-3b44-45fd-a13a-309f592999f5" />

4. Prettify along with Format: 
<img width="1280" alt="Screenshot 2025-05-26 at 3 10 56 AM" src="https://github.com/user-attachments/assets/a549f640-39b1-41db-af32-08ea46452476" />

6. New line, Backslash, Quotation, LineBreak, tab, Carriage return + newline

|Raw Log|Prettified Log|
|----|----|
|<img width="1237" alt="Screenshot 2025-05-25 at 4 19 24 PM" src="https://github.com/user-attachments/assets/b0862644-a24d-4317-8478-dd68e66643e6" />|<img width="1231" alt="Screenshot 2025-05-25 at 4 19 39 PM" src="https://github.com/user-attachments/assets/1302d6bc-6d60-454e-a118-b2ff3175aece" />|

<details>
<summary>Logs in different Themes</summary>
1. Dark
<img width="1439" alt="Screenshot 2025-05-25 at 4 22 54 PM" src="https://github.com/user-attachments/assets/4f7e69c6-e555-4343-84df-12fd4f4a8a54" />
2. Light
<img width="1439" alt="Screenshot 2025-05-25 at 4 24 23 PM" src="https://github.com/user-attachments/assets/d042cac8-fc3e-41a7-a77a-f6c87c046e45" />
3. Headlamp Classic 
<img width="1439" alt="Screenshot 2025-05-25 at 4 25 16 PM" src="https://github.com/user-attachments/assets/46f9c053-76c5-4819-9f28-97e22e16fd62" />
4. Lights out
<img width="1439" alt="Screenshot 2025-05-25 at 4 25 48 PM" src="https://github.com/user-attachments/assets/b9e4431e-35da-4756-afc4-211d96c66e8c" />
5. Monochrome Light
<img width="1439" alt="Screenshot 2025-05-25 at 4 26 26 PM" src="https://github.com/user-attachments/assets/d698327f-8101-44eb-8bcc-fa7752731849" />

</details>

<details>
<summary>Story State of logs</summary>
1. Plain Logs
<img width="1440" alt="Screenshot 2025-05-26 at 4 32 18 AM" src="https://github.com/user-attachments/assets/3d54a209-f6eb-4382-9439-8415a9b49df1" />

2. Json Logs
<img width="1440" alt="Screenshot 2025-05-26 at 4 32 35 AM" src="https://github.com/user-attachments/assets/3adab3bb-a163-449d-9dbe-c809a0006b57" />

3. Formatted Json Logs
<img width="1440" alt="Screenshot 2025-05-26 at 4 33 10 AM" src="https://github.com/user-attachments/assets/bf74cf3c-ec60-4943-a1dc-5c1325fe0cfb" />

4. Big Json Logs
<img width="1440" alt="Screenshot 2025-05-26 at 4 33 26 AM" src="https://github.com/user-attachments/assets/c986dbb2-a487-4f3f-843a-748080dc95e6" />

5. Formatting Logs
<img width="1440" alt="Screenshot 2025-05-26 at 4 24 52 AM" src="https://github.com/user-attachments/assets/2afac2ea-b093-473b-9f4e-226e0f858913" />


</details>

Fixes #3208

userAdityaa commented:
**Requesting Confirmation on Type Change & Log Handling Behavior**

* Replaced the generic `StreamResultsCb` with `LogStreamResultsCb = (result: { logs: string[]; hasJsonLogs: boolean }) => void;` it works fine from my end, but I’d appreciate a review as I’m a bit unsure.
* Also, please verify how the log handling behaves on your end to ensure everything works as expected.

I will implement the required changes based on the review.

illume commented:
Very cool.

@feym78 what do you think?

illume commented on frontend/src/lib/k8s/pod.ts:
```diff
@@ -162,19 +167,70 @@ class Pod extends KubeObject<KubePod> {
       url += `&tailLines=${tailLines}`;
     }
 
+    const prettifyLogLine = (logLine: string): string => {
```
Can these be functions instead of arrow functions?

userAdityaa replied:
Yes, I have changed it to function now.

illume commented on frontend/src/lib/k8s/pod.ts:
```diff
@@ -256,14 +311,6 @@ class Pod extends KubeObject<KubePod> {
       return this.detailedStatusCache.details;
     }
 
-    // We cache this data to avoid going through all this logic when nothing has changed
```
Can you please explain why this was removed? It seems like something that is useful.

userAdityaa replied:
Errr....I mistakenly removed it. Now it is added again.

illume commented on frontend/src/components/pod/Details.tsx:
```diff
@@ -74,14 +76,45 @@ export function PodLogViewer(props: PodLogViewerProps) {
+    const ANSI_GREEN = '\x1b[32m';
+    const ANSI_RESET = '\x1b[0m';
+
+    const colorizePrettifiedLog = (logEntry: string): string => {
```
I'm not sure how well this works. Would you mind making this an exported function, and creating a few tests for it?

Something like this... (untested) but slightly different I guess. It think it might be worth putting some of these functions into a jsonHandling.ts module? Then just using them inside Details.tsx. Maybe even `prettifyLogLine`?

```ts
import { colorizePrettifiedLog, ANSI_BLUE, ANSI_GREEN ,  ANSI_RESET   } from './Details';

describe('colorizePrettifiedLog', () => {
  it('colors JSON keys and string values', () => {
    const input = `{\n  "name": "Alice"\n}`;
    const expected = `{\n  ${ANSI_BLUE}"name"${ANSI_RESET}: ${ANSI_GREEN}"Alice"${ANSI_RESET}\n}`;
    expect(colorizePrettifiedLog(input)).toBe(expected);
  });

  it('colors numeric values', () => {
    const input = `{\n  "age": 30\n}`;
    const expected = `{\n  ${ANSI_BLUE}"age"${ANSI_RESET}: ${ANSI_GREEN}30${ANSI_RESET}\n}`;
    expect(colorizePrettifiedLog(input)).toBe(expected);
  });

  it('colors boolean and null values', () => {
    const input = `{\n  "active": true,\n  "deleted": null\n}`;
    const expected = `{\n  ${ANSI_BLUE}"active"${ANSI_RESET}: ${ANSI_GREEN}true${ANSI_RESET},\n  ${ANSI_BLUE}"deleted"${ANSI_RESET}: ${ANSI_GREEN}null${ANSI_RESET}\n}`;
    expect(colorizePrettifiedLog(input)).toBe(expected);
  });

  it('handles floats correctly', () => {
    const input = `{\n  "score": 98.6\n}`;
    const expected = `{\n  ${ANSI_BLUE}"score"${ANSI_RESET}: ${ANSI_GREEN}98.6${ANSI_RESET}\n}`;
    expect(colorizePrettifiedLog(input)).toBe(expected);
  });

  it('returns original string on error', () => {
    const input = `not a json string`;
    expect(colorizePrettifiedLog(input)).toBe(input);
  });
});
```

userAdityaa replied:
I have seperated the logic in `jsonHandling.ts` and made the test for it `jsonHandling.test.ts`. If we need any further testing then I will implement that.

illume commented on frontend/src/components/pod/Details.tsx:
```diff
@@ -74,14 +76,45 @@ export function PodLogViewer(props: PodLogViewerProps) {
+    hasJsonLogs: boolean;
+  }) {
+    setHasJsonLogs(hasJsonLogs);
+    const ANSI_BLUE = '\x1b[34m';
```
We would need to check these colors look with dark/light themes.

illume reviewed (changes requested):
I had a chance to review the code a bit, and left some questions and notes.

userAdityaa commented on frontend/src/lib/k8s/pod.ts:
```diff
@@ -162,19 +167,70 @@ class Pod extends KubeObject<KubePod> {
 
+    const prettifyLogLine = (logLine: string): string => {
+      try {
+        const trimmedLog = logLine.trim();
```
Yes, that was indeed redundant I have remove it.

userAdityaa commented on frontend/src/lib/k8s/pod.ts:
```diff
@@ -162,19 +167,70 @@ class Pod extends KubeObject<KubePod> {
+
+        if (showTimestamps) {
+          const timestamp = logLine.substring(0, logLine.indexOf('{')).trim();
+          return timestamp ? `${timestamp}\n${prettyJson}\n` : prettyJson;
```
I have changed it look something like this:
```
const timestamp = logLine.slice(0, jsonMatch.index).trim();
return timestamp ? `${timestamp}\n${prettyJson}\n` : `${prettyJson}\n`;
```

userAdityaa commented on frontend/src/lib/k8s/pod.ts:
```diff
@@ -162,19 +167,70 @@ class Pod extends KubeObject<KubePod> {
+        const prettyJson = JSON.stringify(jsonObj, null, 2);
+
+        if (showTimestamps) {
+          const timestamp = logLine.substring(0, logLine.indexOf('{')).trim();
```
I have changed it: 

```
const timestamp = logLine.slice(0, jsonMatch.index).trim();
```

userAdityaa commented on frontend/src/lib/k8s/pod.ts:
```diff
@@ -162,19 +167,70 @@ class Pod extends KubeObject<KubePod> {
+
+      // Skip non-JSON logs when prettifyLogs is true
+      if (prettifyLogs && !isValidLogLine(decodedLog)) {
+        return; // Discard non-JSON logs
```
I have added this line, cause most of the time because of string coming in chunk the logs look something like this: 
<img width="1232" alt="Screenshot 2025-05-24 at 3 32 27 PM" src="https://github.com/user-attachments/assets/b7a42c5c-4804-4393-8246-efd0f67c069f" />


But I have changed it, if we have other plan then I can implement that.

userAdityaa commented on frontend/src/lib/k8s/pod.ts:
```diff
@@ -162,19 +167,70 @@ class Pod extends KubeObject<KubePod> {
 
-      logs.push(Base64.decode(item));
-      onLogs(logs);
+      const isJson = isValidLogLine(decodedLog);
```
I have removed it, but this is still required: 
```
if (!jsonMatch) return false;
JSON.parse(jsonMatch[1]);
hasJsonLogs = true;
const isJson = isValidLogLine(decodedLog);
```

for showing the Json toggle option as per logs.

feym78 commented:
> Very cool.
> 
> @feym78 what do you think?

Hi.
Definitely like it!

@userAdityaa Just to be clear - there is no new line trailing support yet?

userAdityaa commented:
Hey @feym78, I'm a bit unclear about what you mean by the `newline trailing support`. After the change, the logs currently look like this:
<img width="1011" alt="Screenshot 2025-05-24 at 5 06 23 PM" src="https://github.com/user-attachments/assets/fd55b7c7-f853-482d-9786-9a0a3ab75352" />
And what is the expected behaviour you want ?

feym78 commented:
@userAdityaa I meant behavior discussed here https://github.com/kubernetes-sigs/headlamp/issues/3208#issuecomment-2894082241
Let's say "message" json element have a text like on the first screen from link above, is it possible to add to your PR possibility to treat new line characters so text from prettified "message" element (or any else) will look like on the second screen?

illume commented:
@userAdityaa thanks for all those fixes and improvements! I marked all those notes as resolved.

@feym78 I guess all things like `\r\n \"\t` and more should just be printed out? So if the value is a string, then print it out?

userAdityaa commented:
> @userAdityaa thanks for all those fixes and improvements! I marked all those notes as resolved.
> 
> @feym78
> 
> * I guess all things like `\r\n \"\t` and more should just be printed out?
> * So if the value is a string, then print it out?
> * This would be for everything, not just valid JSON?

I had the same thought and would suggest keeping the formatting as-is. 
Also, should I create atomic commits for the tests, or just leave them as they are?

illume commented:
Sorry, I think prettify on should print \n as actual new lines. Like is shown in the comment https://github.com/kubernetes-sigs/headlamp/issues/3208#issuecomment-2894082241

userAdityaa commented:
I have now added the feature for new line trailing and also updated the logs for different themes in the PR description.

userAdityaa commented:
> > Awesome! Thanks.
> > Regarding previous discussion I'm thinking should line trailing also be optional same as prettify? I mean some switch for it.
> > Looks like not everyone will want to work it this way by default.
> > Or may be add switch, make it enabled by default if prettify is enabled, but with possibility to disable if user doesn't want it this >way or if there are some problems with displaying logs with this feature.
> 
> Interesting point.
> 
> > should line trailing also be optional same as prettify?
> 
> Hrmm. We have lots of switches already though. If we can come up with a solution that works well without a switch it would be best. We could try and get some wider feedback. See it if people like it one way or another?
> 
> What do you think @userAdityaa ?
> 
> btw. It might be useful to have a bunch of different types of logs as storybook states. That would make it easier to see how the logs are being presented with prettify on.

Yes, that approach makes sense. I think we should go with a Grid Action Button, similar to what we use for `Find`, `Clear`, and `Download`, especially if we want to save space for the toggle. We can show a tooltip on hover to indicate its function. Initially, the button won’t have a highlighted background to indicate it’s disabled, and after it's clicked, the background will highlight to show it’s active. The only challenge is deciding which icon would best represent this action.

Before Enabling (Now Download icon) : 
<img width="48" alt="Screenshot 2025-05-25 at 7 36 54 PM" src="https://github.com/user-attachments/assets/efa1f50a-449d-4c9c-9b5d-ca2f603e1355" />

After Enabling (Now Download icon): 
<img width="47" alt="Screenshot 2025-05-25 at 7 37 29 PM" src="https://github.com/user-attachments/assets/ddd3ee2e-04c6-41da-8b7e-b99f2e9ce30a" />

userAdityaa commented:
I have added the `Format` button with a Light tooltip of what it does. And also updated the storybook with different state logs like: Plain, Json, Formatted, Formatting, BigJson. 

I have also updated the screenshots with the format button.
